File: quick_sort_step.java
import utils.CsvReader;
import utils.CsvReader.DataPair;
import java.util.List;
import java.util.ArrayList;
import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Scanner;

public class quick_sort_step extends CsvReader {
    
    public static void quickSort(List<DataPair> data, int low, int high, BufferedWriter writer) throws IOException {
        if (low < high) {
            int pi = partition(data, low, high);

            // Print full snapshot after partition
            writer.write("pi=" + pi + " " + data);
            writer.newLine();

            // Recur on left and right partitions
            quickSort(data, low, pi - 1, writer);
            quickSort(data, pi + 1, high, writer);
        }
    }

    public static int partition(List<DataPair> data, int low, int high) {
        DataPair pivot = data.get(high);
        int i = low - 1;

        for (int j = low; j < high; j++) {
            if (data.get(j).id <= pivot.id) {
                i++;
                // Manual swap instead of Collections.swap()
                DataPair temp = data.get(i);
                data.set(i, data.get(j));
                data.set(j, temp);
            }
        }
        
        // Manual swap instead of Collections.swap()
        DataPair temp = data.get(i + 1);
        data.set(i + 1, data.get(high));
        data.set(high, temp);
        
        return i + 1; // New pivot index
    }

    public static void main(String[] args) {
        CsvReader reader = new CsvReader();
        int startingRow = 0;
        int endingRow = 0;
        
        List<DataPair> data = null;
        List<DataPair> aux = null;
        
        try (Scanner scanner = new Scanner(System.in)) {
            System.out.print("Enter the path to the CSV file: ");
            String csvFile = scanner.nextLine();
            data = reader.readCsv(csvFile);
            
            System.out.print("Enter the starting row: ");
            startingRow = scanner.nextInt();
            System.out.print("Enter the ending row: ");
            endingRow = scanner.nextInt();
        } catch (IOException e) {
            System.err.println("Error reading file: " + e.getMessage());
        }
        
        if (data != null) {
            // Extract subset of data
            aux = new ArrayList<>();
            for (int i = startingRow - 1; i < endingRow && i < data.size(); i++) {
                aux.add(data.get(i));
            }
            
            String filename = "quick_sort_step_" + startingRow + "_" + endingRow + ".txt";
            
            try (BufferedWriter writer = new BufferedWriter(new FileWriter(filename))) {
                // Write initial state
                writer.write(aux.toString());
                writer.newLine();
                
                // Perform quick sort with steps
                quickSort(aux, 0, aux.size() - 1, writer);
                
                System.out.println("Quick sort steps saved to: " + filename);
            } catch (IOException e) {
                e.printStackTrace();
            }
        } else {
            System.out.println("No data to display.");
        }
    }
}

File: quick_sort_step.py
import csv
from typing import List, Tuple

class CsvReader:
    class DataPair:
        def __init__(self, key: str, value: str):
            self.id = int(key)
            self.value = value

        def __lt__(self, other):
            return self.id < other.id

        def __le__(self, other):
            return self.id <= other.id

        def __repr__(self):
            return f"{self.id}/{self.value}"

    def read_csv_with_start_and_end(self, filepath: str, start: int, end: int):
        data = []
        with open(filepath, newline='', encoding='utf-8') as csvfile:
            reader = csv.reader(csvfile)
            all_rows = list(reader)
            start_idx = start - 1
            end_idx = end

            for i in range(start_idx, min(end_idx, len(all_rows))):
                row = all_rows[i]
                if len(row) >= 2:
                    data.append(self.DataPair(row[0], row[1]))
        return data


def quick_sort(data: List[CsvReader.DataPair], low: int, high: int, writer):
    if low < high:
        pi = partition(data, low, high)

        # Print full snapshot after partition
        writer.write(f"pi={pi} {data}\n")

        quick_sort(data, low, pi - 1, writer)
        quick_sort(data, pi + 1, high, writer)


def partition(data: List[CsvReader.DataPair], low: int, high: int) -> int:
    pivot = data[high]  # ✅ Last element as pivot
    i = low - 1

    for j in range(low, high):
        if data[j].id <= pivot.id:  # ✅ Compare .id values (integers)
            i += 1
            data[i], data[j] = data[j], data[i]

    data[i + 1], data[high] = data[high], data[i + 1]
    return i + 1


def main():
    reader = CsvReader()
    csv_file = input("Enter the path to the CSV file: ")
    starting_row = int(input("Enter the starting row: "))
    ending_row = int(input("Enter the ending row: "))

    try:
        data = reader.read_csv_with_start_and_end(csv_file, starting_row, ending_row)
        aux = list(data)  # Unused but included as per original
    except Exception as e:
        print(f"Error reading file: {e}")
        return

    if data:
        filename = f"python_quick_sort_step_{starting_row}_{ending_row}.txt"
        try:
            with open(filename, 'w', encoding='utf-8') as writer:
                writer.write(f"{data}\n")  # initial state
                quick_sort(data, 0, len(data) - 1, writer)
        except Exception as e:
            print(f"Error writing to file: {e}")
    else:
        print("No data to display.")


if __name__ == "__main__":
    main()
